content/client-lib-development-guide/features.textile

```diff
@@ -287,9 +292,10 @@ h3(#realtime-connection). Connection
 * @(RTN6)@ A @Connection@ is successful and considered @CONNECTED@ once the websocket connection is open and the intial @CONNECTED@ @ProtocolMessage@ has been received
 * @(RTN19)@ If the @CONNECTED@ @ProtocolMessage@ contains a @connectionDetails@ property, the attributes within @ConnectionDetails@ will be used as the defaults for this client library, overriding any configured options at the time the @CONNECTED@ @ProtocolMessage@ is received
 * @(RTN7)@ @ACK@ and @NACK@:
-** @(RTN7a)@ All @ProtocolMessage@ @Presence@ and @Message@ objects sent to Ably expect either an @ACK@ or @NACK@ from Ably to confirm successful receipt and acceptance or failure respectively
+** @(RTN7a)@ All @ProtocolMessage@ @Presence@ and @Message@ objects sent to Ably expect either an @ACK@ or @NACK@ from Ably to confirm successful receipt and acceptance or failure respectively. For clarity, it is unnecessary to fail the publish operation of a message using a timer. Instead the client library can rely on the realtime system to send an @ACK@ or @NACK@ when connected, to fail all awaiting messages when @DISCONNECTED@ for too long as described in "RTN7d":#RTN7d, or upon reconnecting, the corresponding @ACK@ or @NACK@ protocol messages will be received as described in "RTN19a":#RTN19a
 ** @(RTN7b)@ Every @ProtocolMessage@ sent must contain a unique serially incrementing @msgSerial@ integer value starting at zero. The @msgSerial@ along with the @count@ for incoming @ACK@ and @NACK@ @ProtocolMessages@ indicates which messages succeeded or failed to be delivered
 ** @(RTN7c)@ If a connection enters the @CLOSED@ or @FAILED@ state, or if the connection state is lost, and a @ACK@ or @NACK@ has not yet been received for a message, the client should consider the delivery of those messages as failed
+** @(RTN7d)@ If a connection enters the @DISCONNECTED@ or @SUSPENDED@ state, and has not become @CONNECTED@ within 15 seconds, then the client should consider the delivery of those messages as failed
```

Comment:
>> he messages awaiting ACK/NACK may have in fact been sent and if the client reconnects we will find out

> Unless I'm misunderstanding, I don't think we will - when it reconnects it'll be a different transport, so no acks/nacks from the previous. We'll never know. As I understand it, what we can do instead is resend them with the original msgSerial, and realtime will dedup if it already received a msg with that serial (though ably-js does not do that at the moment).

Yes, we will never know, but the realtime system will send an ACK or NACK back when the client library publishes the message again on the new transport.  So if we failed all messages already sent to the realtime system, then it's quite plausible devs will be told their message send failed and will send again, only to discover it was published twice.  By resending the original messages on the new transport the realtime system prevents that from happening.  

> @paddybyers to be clear, would you say we should fail all queued messages after 15s of disconnection then (whether or not a send has previously been attempted for that message)?

We have never stated that we resend queued messages, we can do that, but I don't think it's right personally. For example take this scenario:

* User is connected and attached
* User publishes 3 messages 
* Connection becomes disconnected
* User publishes 3 more messages
* Connection becomes disconnected
* First 3 messages auto re-published and next 3 messages that are queued are sent.  ACKs are received for all 6 even though the first 3 messages got through in the old connection.